### PR TITLE
FrameTiming build start timestamp fix and add vsync start timestamp

### PR DESCRIPTION
**Pull request by CareF**
## Description

See flutter/flutter#62689
The original build_start time stamp is now vsync_start, and the current build time closer to what we get from timeline events. 

## Related Issues

fixes flutter/flutter#62689

## Tests

I modify the `'FrameTiming.toString has the correct format'` test in `testing/dart/window_test.dart`

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: flow/layers/layer_tree.h [hunk 1]
```diff
@@ -50,7 +50,12 @@ class LayerTree {
   const SkISize& frame_size() const { return frame_size_; }
   float device_pixel_ratio() const { return device_pixel_ratio_; }
 
-  void RecordBuildTime(fml::TimePoint build_start, fml::TimePoint target_time);
+  void RecordBuildTime(fml::TimePoint vsync_start,
+                       fml::TimePoint build_start,
+                       fml::TimePoint target_time);
+  fml::TimePoint vsync_start() const { return vsync_start_; }
+  fml::TimeDelta vsync_overhead() const { return build_start_ - vsync_start_; }
+  fml::TimeDelta vsync_time() const { return build_finish_ - vsync_start_; }
```
Comment (iskakaushik) [inline flow/layers/layer_tree.h:58]: `vsync_time` this isn't accurate. The name suggests that this is the time between adjacent vsync signals, which this isn't. Why do we need this method?

File: shell/common/animator.cc [hunk 2]
```diff
@@ -133,7 +134,8 @@ void Animator::BeginFrame(fml::TimePoint frame_start_time,
   // to service potential frame.
   FML_DCHECK(producer_continuation_);
 
-  last_frame_begin_time_ = frame_start_time;
+  last_frame_begin_time_ = fml::TimePoint::Now();
```
Comment (iskakaushik) [inline shell/common/animator.cc:137]: when we post a trace event for `VsyncSchedulingOverhead` we also use `fml::TimePoint::Now()`. Can we unify both these and maybe post a trace event here instead?

File: flow/layers/layer_tree.h (hunk 1 above)
Comment (CareF) [reply flow/layers/layer_tree.h]: Removed it.

File: shell/common/animator.cc (hunk 2 above)
Comment (CareF) [reply shell/common/animator.cc:137]: Not 100% sure I understand but I made a commit moving the TaceEventAsyncComplete line. PTAL.

File: flow/layers/layer_tree.h (hunk 1 above)
Comment (CareF) [reply flow/layers/layer_tree.h]: I shouldn't have removed it..

File: shell/common/rasterizer.cc [hunk 3]
```diff
@@ -382,7 +383,7 @@ RasterStatus Rasterizer::DrawToSurface(flutter::LayerTree& layer_tree) {
   // There is no way for the compositor to know how long the layer tree
   // construction took. Fortunately, the layer tree does. Grab that time
   // for instrumentation.
-  compositor_context_->ui_time().SetLapTime(layer_tree.build_time());
+  compositor_context_->ui_time().SetLapTime(layer_tree.vsync_time());
```
Comment (CareF) [inline shell/common/rasterizer.cc:386]: @iskakaushik  the `vsync_time` is needed here. I'm not sure what this `SetLapTime` is doing but by using this we are sending in the same value as the original version.

Comment (iskakaushik) [reply shell/common/rasterizer.cc:386]: Looking at the comment on the top, seems like we should pass the frame build time and remove vsync_time.

Comment (CareF) [reply shell/common/rasterizer.cc:386]: OK. And it sounds like there will be bunch of benchmark we need to rebase?

File: flow/layers/layer_tree.h (hunk 1 above)
Comment (CareF) [reply flow/layers/layer_tree.h]: So remove it again.

File: shell/common/rasterizer.cc (hunk 3 above)
Comment (CareF) [reply shell/common/rasterizer.cc]: Removed `vsync_time`.

File: testing/dart/window_test.dart [hunk 4]
```diff
@@ -22,8 +22,8 @@ void main() {
   });
 
   test('FrameTiming.toString has the correct format', () {
-    final FrameTiming timing = FrameTiming(<int>[1000, 8000, 9000, 19500]);
-    expect(timing.toString(), 'FrameTiming(buildDuration: 7.0ms, rasterDuration: 10.5ms, totalSpan: 18.5ms)');
+    final FrameTiming timing = FrameTiming(<int>[500, 1000, 8000, 9000, 19500]);
+    expect(timing.toString(), 'FrameTiming(buildDuration: 6.5ms, rasterDuration: 10.5ms, vsyncOverhead: 0.5ms, totalSpan: 18.5ms)');
```
Comment (liyuqian) [inline testing/dart/window_test.dart:26]: I still expect the `buildDuration` to be 7ms. I can also probably understand if it's 7.5ms (8000 - 500). But I'm not sure how this 6.5ms is calculated?

File: lib/ui/window.dart [hunk 5]
```diff
@@ -47,6 +47,11 @@ typedef _SetNeedsReportTimingsFunc = void Function(bool value);
 ///
 /// [FrameTiming] records a timestamp of each phase for performance analysis.
 enum FramePhase {
+  /// When the UI thread receives the vsync signal from the operating system.
```
[the remark this answers is not in the document]
Comment (CareF) [reply lib/ui/window.dart]: Done.

File: lib/ui/window.dart [hunk 6]
```diff
@@ -121,22 +126,26 @@ class FrameTiming {
   /// {@macro dart.ui.FrameTiming.fps_milliseconds}
   Duration get rasterDuration => _rawDuration(FramePhase.rasterFinish) - _rawDuration(FramePhase.rasterStart);
 
+  /// The duration between receiving the vsync signal and starting building the
+  /// frame.
+  Duration get vsyncOverhead => _rawDuration(FramePhase.vsyncStart) - _rawDuration(FramePhase.buildStart);
```
[the remark this answers is not in the document]
Comment (CareF) [reply lib/ui/window.dart]: Done.

File: lib/web_ui/lib/src/ui/window.dart [hunk 7]
```diff
@@ -1067,6 +1072,10 @@ class FrameTiming {
       _rawDuration(FramePhase.rasterFinish) -
       _rawDuration(FramePhase.rasterStart);
 
+  /// The duration between receiving the vsync signal and starting building the
+  /// frame.
+  Duration get vsyncOverhead => _rawDuration(FramePhase.vsyncStart) - _rawDuration(FramePhase.buildStart);
```
[the remark this answers is not in the document]
Comment (CareF) [reply lib/web_ui/lib/src/ui/window.dart]: Done

File: lib/web_ui/lib/src/ui/window.dart [hunk 8]
```diff
@@ -992,6 +992,11 @@ class IsolateNameServer {
 ///
 /// [FrameTiming] records a timestamp of each phase for performance analysis.
 enum FramePhase {
+  /// When the UI thread receives the vsync signal from the operating system.
```
[the remark this answers is not in the document]
Comment (CareF) [reply lib/web_ui/lib/src/ui/window.dart]: Done.

File: lib/ui/window.dart [hunk 9]
```diff
@@ -121,22 +126,26 @@ class FrameTiming {
...
+
   /// The timespan between build start and raster finish.
   ///
   /// To achieve the lowest latency on an X fps display, this should not exceed
   /// 1000/X milliseconds.
   /// {@macro dart.ui.FrameTiming.fps_milliseconds}
   ///
   /// See also [buildDuration] and [rasterDuration].
-  Duration get totalSpan => _rawDuration(FramePhase.rasterFinish) - _rawDuration(FramePhase.buildStart);
+  Duration get totalSpan => _rawDuration(FramePhase.rasterFinish) - _rawDuration(FramePhase.vsyncStart);
```
[the remark this answers is not in the document]
Comment (CareF) [reply lib/ui/window.dart:140]: I changed the doc.

File: lib/web_ui/lib/src/ui/window.dart [hunk 10]
```diff
@@ -1076,15 +1085,15 @@ class FrameTiming {
   /// See also [buildDuration] and [rasterDuration].
   Duration get totalSpan =>
       _rawDuration(FramePhase.rasterFinish) -
-      _rawDuration(FramePhase.buildStart);
+      _rawDuration(FramePhase.vsyncStart);
```
[the remark this answers is not in the document]
Comment (CareF) [reply lib/web_ui/lib/src/ui/window.dart:1088]: Done.

File: testing/dart/window_test.dart (hunk 4 above)
Comment (CareF) [reply testing/dart/window_test.dart]: Done. should be 7ms.

**Comment by CareF** [comment]
I realized that I need also to update the framework code as in flutter/flutter#62933 . How should this work? @liyuqian

**Comment by liyuqian** [comment]
> I realized that I need also to update the framework code as in [flutter/flutter#62933](https://github.com/flutter/flutter/pull/62933) . How should this work? @liyuqian
> Especially `packages/flutter/test/scheduler/scheduler_test.dart` because this test depends on the `FrameTiming` data length which we modify here. This will block the CI test on the framework.

I'd suggest adding a new `FrameTiming` constructor that takes named parameter for unit tests, and document that the `List<int>` constructor should only be used in the engine as `FramePhase` could change. Here are some specific steps:
1. Introduce that named constructor in an engine PR
2. Modify the unit tests to use the named constructor in a framework PR
3. Land this PR by extending the named constructor without breaking the old unit tests
4. Update the unit test by giving the new named parameter in another framework PR (likely https://github.com/flutter/flutter/pull/62933)

Additionally, we can make the `List<int>` constructor private as `FrameTimng._(List<int>)`. The [hooks.dart] can still access those private constructors, but I think framework won't be able to.

[hooks.dart]: https://github.com/flutter/engine/blob/master/lib/ui/hooks.dart

**Comment by liyuqian** [comment]
Should this PR also update `FrameTiming.fromTimeStamps`, and make the old `List<int>` constructor private?

**Comment by CareF** [comment]
> Should this PR also update `FrameTiming.fromTimeStamps`, and make the old `List<int>` constructor private?

I just duplicate `fromTimeStamps` to default constructor and will remove `fromTimeStamps` after everything lands.

**Pull request merged**